### ISSUE TITLE
Remove redundant lock in z_sendmany call

### ISSUE THREAD

**Issue by bitcartel** [zcash/zcash#2623]
The lock below can be removed.

```
UniValue AsyncRPCOperation_sendmany::perform_joinsplit(AsyncJoinSplitInfo & info) {
    std::vector<boost::optional < ZCIncrementalWitness>> witnesses;
    uint256 anchor;
    {
        LOCK2(cs_main, pwalletMain->cs_wallet);
        anchor = pcoinsTip->GetBestAnchor();    // As there are no inputs, ask the wallet for the best anchor
    }
    return perform_joinsplit(info, witnesses, anchor);
}
```

**Comment by daira** [comment]
Are you sure? I thought that pcoinsTip's best anchor could be updated asynchronously (but I could be wrong).

**Comment by bitcartel** [comment]
@daira See comment here (via @str4d) : https://github.com/zcash/zcash/pull/2615#discussion_r139941276

So pcoinsTip is of class CCoinsViewCache, where the .h and .cpp do not document if the class (in general) is thread safe or not.

I see that GetBestAnchor() is called from a few places, and there are no locks in place for (all of) them.  It should be a read operation, although it _does_ mutate the first time.

@ebfull is it thread safe to call GetBestAnchor() or do we need a lock?

Note that GetBestBlock() which is the same code but with a different member variable, doesn't always have a lock.
https://github.com/zcash/zcash/blob/8214ebc61c72e4c3f6d2dc35c39e7a6521d683f9/src/txdb.cpp#L192
via rpc call https://github.com/zcash/zcash/blob/e0bef1de56170ce072cdba621f46cd01b1c9873d/src/rpcblockchain.cpp#L490

whereas it does here (but maybe related to locking on mapBlockIndex), see comment here 
https://github.com/zcash/zcash/blob/99c4c6de0c966081070dcec8d740039df2a23a0b/src/main.h#L520 and code here https://github.com/zcash/zcash/blob/f34c60de379b4fa514ba1489a876d59410f2bc5b/src/main.cpp#L1623

Whatever we decide for this ticket, should be reflected in z_shieldcoinbase as part of 1.0.12 release.

**Comment by daira** [comment]
Ugh. Remember that reads of a shared variable do need a lock, not just writes. Reading outside the lock is undefined behaviour.

**Comment by bitcartel** [comment]
@str4d I think this is a blocker for 1.0.12 until resolved.

@daira We could go ahead and remove the lock, but add a local lock/mutex for GetBestAnchor().

@ebfull Do you recall what the original intent for the method was?  Thread safe or not?

**Comment by str4d** [comment]
> @daira See comment here (via @str4d)

My comment was about removing the lock on `cs_wallet`, not `cs_main`. I've just gone back and looked at the `z_shieldcoinbase` RPC and seen that both were removed; I think the `cs_main` lock should be re-added. We should review other uses of `pcoinsTip` to check in what circumstances they are locked.

**Comment by bitcartel** [comment]
@str4d Sounds like a good plan.  I've created PR #2632 for both z_sendmany and z_shieldcoinbase.

**Comment by str4d** [comment]
Closed by #2632.
